keps/sig-node/4680-add-resource-health-to-pod-status/README.md

```diff
@@ -0,0 +1,453 @@
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+Device Plugin and DRA are relatively new features and have a reasonable test coverage.
+
+##### Unit tests
+
+- `k8s.io/kubernetes/pkg/kubelet/cm/devicemanager`: `5/31/2024` - `84.1`
+- `k8s.io/kubernetes/pkg/kubelet/cm/dra`: `5/31/2024` - `59.2`
+- `k8s.io/kubernetes/pkg/kubelet/cm/dra/plugin`: `5/31/2024` - `34`
+- `k8s.io/kubernetes/pkg/kubelet/cm/dra/state`: `5/31/2024` - `98`
+
+##### Integration tests
+
+N/A
+
+##### e2e tests
+
+Planned tests:
+
+- Device marked unhealthy - the state is reflected in pod status
+- Device marked unhealthy and back to healthy after some time - pod status was changed to unhealthy temporarily
+- Device marked as unhealthy and back to healthy in quick succession - pod status reflects the latest health status
+- Pod failed due to unhealthy device, earlier than device plugin detected it. Pod status is still updated.
+- Pod is in crash loop backoff due to unhealthy device - pod status is updated to unhealthy
+
+Test coverage will be listed once tests are implemented.
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+#### Alpha
+
+- New field is introduced in Pod Status
+- Feature implemented in Device Manager behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Feature is implemented in DRA behind the feature flag
+- Complete e2e tests coverage
+
+#### GA
+
+- Feedback is collected on usability of the field
+- Example of real-world usage with one of the device plugin. For example, NVIDIA Device Plugin
+
+### Upgrade / Downgrade Strategy
+
+The feature exposes a new field based on information the Device Plugin already exposes. There will be no dependency on upgrade/downgrade, feature will either work or not.
+
+DRA implementation requires DRA interfaces change. DRA is in alpha and in active development. The feature will follow the DRA ugrade/downgrade strategy.
+
+### Version Skew Strategy
+
+There is no issue with the version skew. Kubelet that will expose this flag will
+always be the same version of behind the API, which introduced this new field.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+Simple change of a feature gate will either enable or disable this feature.
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `ResourceHealthStatus`
+  - Components depending on the feature gate: `kubelet` and `kube-apiserver`
+
+###### Does enabling the feature change any default behavior?
+
+No
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, with no side effect except of missing the new field in pod status.
```

Comment:
When the feature gate is disabled any instances of the new field already written to pod status will continue to exist.  The typical behavior in this case is to add field wiping such that when the feature is disabled, the field is dropped from writes that contain the new fields if it is not already set, and any code that reads the new fields ignores it.  Note that this does mean that if the feature is turned off and then turned on again, any resources with the new fields will keep the values from when the feature was first enabled.

Reply:
Added a few words here